File: src/Core/Desktop/Sakuno.ING.Views.Desktop.Common/Controls/SlotAircraftPresenter.cs
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using Sakuno.ING.Game.Models;

namespace Sakuno.ING.Views.Desktop.Controls
{
    public class SlotAircraftPresenter : Control
    {
        public static readonly DependencyProperty AircraftProperty
            = DependencyProperty.Register(nameof(Aircraft), typeof(ClampedValue), typeof(SlotAircraftPresenter),
                new PropertyMetadata(new ClampedValue(), (d, e) => ((SlotAircraftPresenter)d).UpdateCount()));
        public ClampedValue Aircraft
        {
            get => (ClampedValue)GetValue(AircraftProperty);
            set => SetValue(AircraftProperty, value);
        }

        public static readonly DependencyProperty IsPlaneProperty
            = DependencyProperty.Register(nameof(IsPlane), typeof(bool), typeof(SlotAircraftPresenter),
                new PropertyMetadata(false, (d, e) => ((SlotAircraftPresenter)d).CheckVisualState()));
        public bool IsPlane
        {
            get => (bool)GetValue(IsPlaneProperty);
            set => SetValue(IsPlaneProperty, value);
        }

        private void CheckVisualState()
        {
            var a = Aircraft;
            if (!IsPlane)
                textblock.Foreground = Brushes.Gray;
            else if (a.Current == a.Max)
                textblock.Foreground = Brushes.SpringGreen;
            else if (a.Current == 0)
                textblock.Foreground = Brushes.Red;
            else
                textblock.Foreground = Brushes.GreenYellow;
        }

        private void UpdateCount()
        {
            textblock.Text = Aircraft.Current.ToString();
            CheckVisualState();
        }

        private readonly TextBlock textblock = new TextBlock();

        public SlotAircraftPresenter() => AddVisualChild(textblock);
        protected override int VisualChildrenCount => 1;
        protected override Visual GetVisualChild(int index) => textblock;
    }
}

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/SlotAircraftPresenter.vb
Imports Sakuno.ING.Game.Models

Namespace Controls
    <TemplatePart(Name:="PART_Text", Type:=GetType(TextBlock))>
    <TemplateVisualState(GroupName:="State", Name:="NotPlane")>
    <TemplateVisualState(GroupName:="State", Name:="Full")>
    <TemplateVisualState(GroupName:="State", Name:="Empty")>
    <TemplateVisualState(GroupName:="State", Name:="Half")>
    Public NotInheritable Class SlotAircraftPresenter
        Inherits Control

        Private _isPlane As Boolean
        Public Property IsPlane As Boolean
            Get
                Return _isPlane
            End Get
            Set(value As Boolean)
                _isPlane = value
                CheckVisualState()
            End Set
        End Property

        Private _aircraft As ClampedValue
        Public Property Aircraft As ClampedValue
            Get
                Return _aircraft
            End Get
            Set(value As ClampedValue)
                _aircraft = value
                UpdateCount()
            End Set
        End Property

        Public Sub New()
            DefaultStyleKey = GetType(SlotAircraftPresenter)
        End Sub

        Private Sub CheckVisualState()
            If Not IsPlane Then
                VisualStateManager.GoToState(Me, "NotPlane", True)
            ElseIf Aircraft.Current = Aircraft.Max Then
                VisualStateManager.GoToState(Me, "Full", True)
            ElseIf Aircraft.Current = 0 Then
                VisualStateManager.GoToState(Me, "Empty", True)
            Else
                VisualStateManager.GoToState(Me, "Half", True)
            End If
        End Sub

        Private Sub UpdateCount()
            If PART_Text IsNot Nothing Then PART_Text.Text = Aircraft.Current.ToString()
            CheckVisualState()
        End Sub

        Private PART_Text As TextBlock
        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            PART_Text = TryCast(GetTemplateChild("PART_Text"), TextBlock)
            UpdateCount()
        End Sub

    End Class
End Namespace